MDANSE/Src/MDANSE/MolecularDynamics/Trajectory.py

```diff
@@ -558,62 +598,72 @@ def write_atom_properties(
-                    # happen when the trajectory contained custom atom
-                    # properties and this function was inputted with
-                    # ptypes=None. Let's derive the type from the
-                    # properties value.
-                    if isinstance(properties[label], (float, np.float32, np.float64)):
-                        type_dataset[next_index] = "float"
-                    elif isinstance(properties[label], (int, np.int32, np.int64)):
-                        type_dataset[next_index] = "int"
-                    elif isinstance(properties[label], str):
-                        type_dataset[next_index] = "str"
-                    elif isinstance(properties[label], list):
-                        type_dataset[next_index] = "list"
-                    else:
-                        raise TypeError(
-                            f"The trajectories atom database contains atom "
-                            f"properties of an unexpected type: "
-                            f"{type(properties[label])}")
-                next_index += 1
+        punits["dummy"] = "none"
+        really_new_labels = set(new_labels) - set(old_labels)
+        for label in really_new_labels:
+            label_dataset[next_index] = label
+            type_dataset[next_index] = ptypes[label]
+            if label in ["b_coherent", "b_incoherent"]:
+                unit_dataset[next_index] = "ang"
+            else:
+                unit_dataset[next_index] = punits[label]
+            next_index += 1
         mapping = {
             property_label.decode("utf-8"): index
             for index, property_label in enumerate(label_dataset[:])
         }
-        atom_dataset = group.create_dataset(symbol, data=200 * [-1.0])
+        atom_dataset = group.create_dataset(symbol, data=200 * [-1.0], dtype=complex)
         for key, value in properties.items():
             try:
-                float(value)
+                numval = complex(value)
             except ValueError:
                 continue
             except TypeError:
                 continue
             else:
-                atom_dataset[mapping[key]] = value
+                atom_dataset[mapping[key]] = numval
+        colour = properties["color"]
         try:
-            colour = [int(x) for x in properties["color"].split(";")]
-        except AttributeError:
-            colour = [int(x) for x in properties["color"][0].split(";")]
-        atom_dataset[mapping["color"]] = (
-            0x10000 * colour[0] + 0x100 * colour[1] + colour[2]
-        )
+            temp = int(colour)
+        except (TypeError, ValueError):
+            try:
+                colour = [int(x) for x in colour.split(";")]
+            except AttributeError:
+                colour = [int(x) for x in colour[0].split(";")]
+            atom_dataset[mapping["color"]] = (
+                0x10000 * colour[0] + 0x100 * colour[1] + colour[2]
+            )
```

Comment:
This strikes me as much clearer with fewer magic numbers
```suggestion
                colour = bytes(map(int, colour.split(";")))
            except AttributeError:
                colour = bytes(map(int, colour[0].split(";")))
            atom_dataset[mapping["color"]] = int.from_bytes(colour, byteorder="big")
```